clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -252,6 +254,82 @@ private void maybeUpdateJoinedSubscription(Set<TopicPartition> assignedPartition
         }
     }
 
+    private Exception maybeAssignPartitions(final Set<TopicPartition> assignedPartitions) {
+        if (!assignedPartitions.isEmpty()) {
+            log.info("Setting newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
+
+            ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
+            try {
+                listener.onPartitionsAssigned(assignedPartitions);
+            } catch (WakeupException | InterruptException e) {
+                throw e;
+            } catch (Exception e) {
+                log.error("User provided listener {} failed on partition assignment of {}",
+                    listener.getClass().getName(), assignedPartitions, e);
+                return e;
+            }
+        }
+
+        return null;
+    }
+
+    private Exception maybeRevokePartitions(final Predicate<TopicPartition> predicate) {
+        Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
```

Comment:
This shouldn't be initialized to assignedPartitions, in onJoinComplete we should be revoking the set Minus(owned-partitions, assigned-partitions) -- this is revoking Minus(assigned-partitions, owned-partitions) Also we are mixing up the naming between owned/assigned which are important to differentiate.
(We can't just hardcode initializing this to ownedPartitions because we need to call revoke on assignedPartitions in onJoinPrepare. Maybe we could pass in the "starting set" ie ownedPartitions or assignedPartitions as a parameter?)

Reply:
Yup.